# Challenge 2

In this challenge we will continue working with the `Pokemon` dataset. We will attempt solving a slightly more complex problem in which we will practice the iterative data analysis process you leaned in [this video](https://www.youtube.com/watch?v=xOomNicqbkk).

The problem statement is as follows:

**You are at a Pokemon black market planning to buy a Pokemon for battle. All Pokemon are sold at the same price and you can only afford to buy one. You cannot choose which specific Pokemon to buy. However, you can specify the type of the Pokemon - one type that exists in either `Type 1` or `Type 2`. Which type should you choose in order to maximize your chance of receiving a good Pokemon?**

To remind you about the 3 steps of iterative data analysis, they are:

1. Setting Expectations
1. Collecting Information
1. Reacting to Data / Revising Expectations

Following the iterative process, we'll guide you in completing the challenge.


## Problem Solving Iteration 1

In this iteration we'll analyze the problem and identify the breakthrough. The original question statement is kind of vague because we don't know what a *good pokemon* really means as represented in the data. We'll start by understanding the dataset and see if we can find some insights.

In [2]:
# Import libraries
import numpy as np
import pandas as pd

In [3]:
pokemon = pd.read_csv("Pokemon.csv")

From the data it seems whether a pokemon is good depends on its abilities as represented in the fields of `HP`, `Attack`, `Defense`, `Sp. Atk`, `Sp. Def`, `Speed`, and `Total`. We are not sure about `Generation` and `Legendary` because they are not necessarily the decisive factors of the pokemon abilities.

But `HP`, `Attack`, `Defense`, `Sp. Atk`, `Sp. Def`, `Speed`, and `Total` are a lot of fields! If we look at them all at once it's very complicated. This isn't Mission Impossible but it's ideal that we tackle this kind of problem after we learn Machine Learning (which you will do in Module 3). For now, is there a way to consolidate the fields we need to look into?

Fortunately there seems to be a way. It appears the `Total` field is computed based on the other 6 fields. But we need to prove our theory. If we can approve there is a formula to compute `Total` based on the other 6 abilities, we only need to look into `Total`.

We have the following expectation now:

#### The `Total` field is computed based on `HP`, `Attack`, `Defense`, `Sp. Atk`, `Sp. Def`, and `Speed`.

We need to collect the following information:

* **What is the formula to compute `Total`?**
* **Does the formula work for all pokemon?**

In the cell below, make a hypothesis on how `Total` is computed and test your hypothesis.

In [7]:
#For the first hipotesis let's assume that total is the sum of the other 6 fields and try it.

# Calculate the sum of HP, Attack, Defense, Sp. Atk, Sp. Def, and Speed
sum_total= pokemon['HP'] + pokemon['Attack'] + pokemon['Defense'] + pokemon['Sp. Atk'] + pokemon['Sp. Def'] + pokemon['Speed']

# Let's compare sum_total to the Total field
matches_total = sum_total  == pokemon['Total']

# Check if the hypothesis is correct
hypothesis_true = matches_total.all()

hypothesis_true

True

# What is the formula to compute Total?
The formula to compute Total is just the sum of HP, Attack, Defense, Sp. Atk, Sp. Def and Speed, as seen above on sum_total.
# Does the formula work for all pokemon?
Yes, it works as shown above.

## Problem Solving Iteration 2

Now that we have consolidated the abilities fields, we can update the problem statement. The new problem statement is:

### Which pokemon type is most likely to have the highest `Total` value?

In the updated problem statement, we assume there is a certain relationship between the `Total` and the pokemon type. But we have two *type* fields (`Type 1` and `Type 2`) that have string values. In data analysis, string fields have to be transformed to numerical format in order to be analyzed. 

In addition, keep in mind that `Type 1` always has a value but `Type 2` is sometimes empty (having the `NaN` value). Also, the pokemon type we choose may be either in `Type 1` or `Type 2`.

Now our expectation is:

#### `Type 1` and `Type 2` string variables need to be converted to numerical variables in order to identify the relationship between `Total` and the pokemon type.

The information we need to collect is:

#### How to convert two string variables to numerical?

Let's address the first question first. You can use a method called **One Hot Encoding** which is frequently used in machine learning to encode categorical string variables to numerical. The idea is to gather all the possible string values in a categorical field and create a numerical field for each unique string value. Each of those numerical fields uses `1` and `0` to indicate whether the data record has the corresponding categorical value. A detailed explanation of One Hot Encoding can be found in [this article](https://hackernoon.com/what-is-one-hot-encoding-why-and-when-do-you-have-to-use-it-e3c6186d008f). You will formally learn it in Module 3.

For instance, if a pokemon has `Type 1` as `Poison` and `Type 2` as `Fire`, then its `Poison` and `Fire` fields are `1` whereas all other fields are `0`. If a pokemon has `Type 1` as `Water` and `Type 2` as `NaN`, then its `Water` field is `1` whereas all other fields are `0`.

#### In the next cell, use One Hot Encoding to encode `Type 1` and `Type 2`. Use the pokemon type values as the names of the numerical fields you create.

The new numerical variables you create should look like below:

![One Hot Encoding](../images/one-hot-encoding.png)

In [11]:
# encode Type 1 and Type 2 with panda (get_dummies). Should i import sklearn.preprocessing from OneHotEncoder isntead?
type_1_encoded = pd.get_dummies(pokemon['Type 1'], prefix='Type 1')
type_2_encoded = pd.get_dummies(pokemon['Type 2'], prefix='Type 2')

# concat the encoded columns
encoded_poke =pd.concat([pokemon, type_1_encoded, type_2_encoded], axis=1)

## Problem Solving Iteration 3

Now we have encoded the pokemon types, we will identify the relationship between `Total` and the encoded fields. Our expectation is:

#### There are relationships between `Total` and the encoded pokemon type variables and we need to identify the correlations.

The information we need to collect is:

#### How to identify the relationship between `Total` and the encoded pokemon type fields?

There are multiple ways to answer this question. The easiest way is to use correlation. In the cell below, calculate the correlation of `Total` to each of the encoded fields. Rank the correlations and identify the #1 pokemon type that is most likely to have the highest `Total`.

In [13]:
correlations = encoded_poke.corr()['Total']
correlations

C:\Users\jmcso\AppData\Local\Temp\ipykernel_6572\432871535.py:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  correlations = encoded_poke.corr()['Total']


#                  0.119813
Total              1.000000
HP                 0.618748
Attack             0.736211
Defense            0.612787
Sp. Atk            0.747250
Sp. Def            0.717609
Speed              0.575943
Generation         0.048384
Legendary          0.501758
Type 1_Bug        -0.143957
Type 1_Dark        0.017818
Type 1_Dragon      0.196532
Type 1_Electric    0.016715
Type 1_Fairy      -0.026948
Type 1_Fighting   -0.029086
Type 1_Fire        0.050527
Type 1_Flying      0.029504
Type 1_Ghost       0.007594
Type 1_Grass      -0.036057
Type 1_Ground      0.004082
Type 1_Ice        -0.002412
Type 1_Normal     -0.104150
Type 1_Poison     -0.057123
Type 1_Psychic     0.094364
Type 1_Rock        0.037524
Type 1_Steel       0.082000
Type 1_Water      -0.015640
Type 2_Bug        -0.021375
Type 2_Dark        0.065844
Type 2_Dragon      0.115240
Type 2_Electric    0.014669
Type 2_Fairy      -0.024606
Type 2_Fighting    0.138726
Type 2_Fire        0.073234
Type 2_Flying      0

# Bonus Question

Say now you can choose both `Type 1` and `Type 2` of the pokemon. In order to receive the best pokemon, which types will you choose?

I would choose the Types with max positive correlation.
In this case, we can chec that type 1 Dragon has 0.196532 correlation and Type 2 Fighting has 0.138726, 
So i would choose Type1 Dragon and Type 2 Fighting.
